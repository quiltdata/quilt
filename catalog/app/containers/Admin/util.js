import mapValues from 'lodash/mapValues';
import omit from 'lodash/omit';
import { compose, withHandlers, withProps, withStateHandlers } from 'recompose';

export const formatDate = (d) => d ? new Date(d).toLocaleString() : 'N/A';

export const withStatefulActions = (bindActions) => compose(
  withStateHandlers({ pending: {} }, {
    startAction: ({ pending }) => (key) => ({
      pending: Object.assign({}, pending, { [key]: true }),
    }),
    endAction: ({ pending }) => (key) => ({
      pending: omit(pending, [key]),
    }),
  }),
  withHandlers(({ actions }) =>
    mapValues(actions, (action, key) =>
      // eslint-disable-next-line object-curly-newline
      ({ pending, startAction, endAction, actions: { [key]: callback } }) => (arg) => {
        if (pending[arg]) return;
        startAction(arg);
        callback(arg)
          .then(() => endAction(arg))
          .catch((e) => { endAction(arg); throw e; });
      })),
  withProps((props) => ({ bindActions: bindActions(props) })),
);
