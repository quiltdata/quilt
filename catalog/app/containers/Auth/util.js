import { stringify } from 'query-string';

import { LATENCY_SECONDS } from './constants';

export const makeHeadersFromTokens = ({ token }) => ({
  Authorization: `Bearer ${token}`,
});

const makeURL = (url, query) => {
  const qs = stringify(query);
  const search = qs ? `?${qs}` : '';
  return `${url}${search}`;
};

const getPath = (loc) => loc.pathname + loc.search;

export const makeSignInURL = (next = getPath(window.location)) =>
  // TODO(dima): Sign up vs sign in?
  makeURL('/signin', { next });

export const adjustTokensForLatency = (tokens) => ({
  ...tokens,
  expires_at:
    Number.isFinite(tokens.expires_at)
      ? tokens.expires_at - LATENCY_SECONDS
      : tokens.expires_at,
});
