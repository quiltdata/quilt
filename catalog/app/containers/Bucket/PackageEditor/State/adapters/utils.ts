import type { FileWithPath } from 'react-dropzone'

import { Status } from 'components/FileManager/FileRow'
import type { TreeEntry } from 'components/FileManager/FileTree'
import type * as Model from 'model'
import * as s3paths from 'utils/s3paths'

export function calcChildren(
  entry: Model.PackageEntry | Model.S3File | FileWithPath,
  tailParts: string[],
  children: TreeEntry[] = [],
): TreeEntry[] {
  const [name, ...tail] = tailParts
  const found = children.find((child) => child.id === name)
  if (found) {
    if (tail.length) {
      found.children = calcChildren(entry, tail, found.children)
    }
    return children
  }

  return children.concat({
    id: name,
    name: tail.length ? s3paths.ensureSlash(name) : name,
    size: tail.length ? 0 : entry.size,
    status: Status.Unchanged,
    children: tail.length ? calcChildren(entry, tail, []) : undefined,
  })
}

export function sortEntries(entries: TreeEntry[]): TreeEntry[] {
  return [...entries].sort((a, b) => {
    if (a.children && !b.children) return -1
    if (!a.children && b.children) return 1
    return a.name.localeCompare(b.name)
  })
}
