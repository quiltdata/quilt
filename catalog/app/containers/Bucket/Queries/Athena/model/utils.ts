export const Loading = Symbol('loading')

export type Maybe<T> = T | null

// `T` is the value
// `undefined` is no data. It is not initialized
// `Loading` is loading
// `Error` is error
export type Data<T> = T | undefined | typeof Loading | Error

export interface DataController<T> {
  data: Data<T>
  loadMore: () => void
}

export function wrapData<T>(data: Data<T>, setPrev: (d: T) => void): DataController<T> {
  return {
    data,
    loadMore: () => hasData(data) && setPrev(data),
  }
}

export interface List<T> {
  list: T[]
  next?: string
}

// `T` is the value
// `null` is no value, explicitly set by user
// `undefined` is no value. It is not initialized
// `Loading` is loading
// `Error` is error
export type Value<T> = Maybe<Data<T>>

export interface ValueController<T> {
  value: Value<T>
  setValue: (v: T | null) => void
}

export function wrapValue<T>(
  value: Value<T>,
  setValue: (d: T | null) => void,
): ValueController<T> {
  return {
    value,
    setValue,
  }
}

/** Data is loaded, or value is set to actual value */
export function hasData<T>(value: Value<T>): value is T {
  if (
    value === undefined ||
    value === Loading ||
    value instanceof Error ||
    value === null
  ) {
    return false
  }
  return true
}

/** No value yet: value or data was just initialized */
export function isNone<T>(value: Value<T>): value is undefined {
  return value === undefined
}

/** Data is loading, or value is waiting for data */
export function isLoading<T>(value: Value<T>): value is typeof Loading {
  return value === Loading
}

export function isError<T>(value: Value<T>): value is Error {
  return value instanceof Error
}

/** Value is selected with some or no value, or resolved with error, or data is loaded (successfully or not) */
export function isReady<T>(value: Value<T>): value is T | null | Error {
  if (value === undefined || value === Loading) {
    return false
  }
  return true
}

/** Value is selected with some or no value, or data is loaded successfully */
export function hasValue<T>(value: Value<T>): value is T | null {
  if (value === undefined || value === Loading || value instanceof Error) {
    return false
  }
  return true
}

/** User explicitly set no value */
export function isNoneSelected<T>(value: Value<T>): value is null {
  return value === null
}
