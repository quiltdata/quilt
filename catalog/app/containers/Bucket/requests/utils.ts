import * as R from 'ramda'

export const decodeS3Key = R.pipe(R.replace(/\+/g, ' '), decodeURIComponent)
