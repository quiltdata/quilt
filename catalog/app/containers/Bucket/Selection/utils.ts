import { join } from 'path'

import * as R from 'ramda'

import type * as Model from 'model'
import * as s3paths from 'utils/s3paths'

export interface SelectionItem {
  logicalKey: string
}

export interface ListingSelection {
  [prefixUrl: string]: SelectionItem[]
}

export const EMPTY_MAP: ListingSelection = {}

interface SelectionHandles {
  [prefixUrl: string]: Model.S3.S3ObjectLocation[]
}

const convertIdToHandle = (
  id: string | number,
  parentHandle: Model.S3.S3ObjectLocation,
): Model.S3.S3ObjectLocation => ({
  bucket: parentHandle.bucket,
  key: join(parentHandle.key, id.toString()),
})

export const toHandlesMap = (selection: ListingSelection): SelectionHandles =>
  Object.entries(selection).reduce(
    (memo, [prefixUrl, items]) => ({
      ...memo,
      [prefixUrl]: items.map((item) =>
        convertIdToHandle(item.logicalKey, s3paths.parseS3Url(prefixUrl)),
      ),
    }),
    {} as SelectionHandles,
  )

export const toHandlesList = (selection: ListingSelection): Model.S3.S3ObjectLocation[] =>
  Object.entries(selection).reduce(
    (memo, [prefixUrl, items]) => [
      ...memo,
      ...items.map((item) =>
        convertIdToHandle(item.logicalKey, s3paths.parseS3Url(prefixUrl)),
      ),
    ],
    [] as Model.S3.S3ObjectLocation[],
  )

const mergeWithFiltered =
  (prefix: string, filteredItems: SelectionItem[]) => (allItems: SelectionItem[]) => {
    if (!allItems || !allItems.length) return filteredItems
    const selectionOutsideFilter = allItems.filter(
      (item) => !item.logicalKey.startsWith(prefix),
    )
    const newIds = [...selectionOutsideFilter, ...filteredItems]
    return R.equals(newIds, allItems) ? allItems : newIds // avoids cyclic update
  }

export function merge(
  items: SelectionItem[],
  bucket: string,
  path: string,
  filter?: string,
): (state: ListingSelection) => ListingSelection {
  const prefixUrl = `s3://${bucket}/${path}`
  const lens = R.lensProp<Record<string, SelectionItem[]>>(prefixUrl)
  return filter ? R.over(lens, mergeWithFiltered(filter, items)) : R.set(lens, items)
}

const EmptyKeys: string[] = []

export const getDirectorySelection = (
  selection: ListingSelection,
  bucket: string,
  path: string,
) => selection[`s3://${bucket}/${path}`] || EmptyKeys
