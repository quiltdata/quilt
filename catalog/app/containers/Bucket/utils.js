import * as React from 'react'

import * as AWS from 'utils/AWS'

export const withDownloadUrl = (handle, callback) => (
  <AWS.Signer.Inject>
    {(signer) =>
      callback(
        signer.getSignedS3URL(handle, { ResponseContentDisposition: 'attachment' }),
      )
    }
  </AWS.Signer.Inject>
)
