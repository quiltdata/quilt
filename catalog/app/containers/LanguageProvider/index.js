/*
 *
 * LanguageProvider
 *
 * this component connects the redux state language locale to the
 * IntlProvider component and i18n messages (loaded from `app/translations`)
 */

import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { mapProps, setPropTypes } from 'recompose';
import { createSelector } from 'reselect';
import { IntlProvider } from 'react-intl';

import { composeComponent } from 'utils/reactTools';
import { injectReducer } from 'utils/ReducerInjector';

import { REDUX_KEY } from './constants';
import reducer from './reducer';
import { makeSelectLocale } from './selectors';

export default composeComponent('LanguageProvider',
  injectReducer(REDUX_KEY, reducer),
  connect(createSelector(
    makeSelectLocale(),
    (locale) => ({ locale })
  )),
  setPropTypes({
    locale: PropTypes.string,
    messages: PropTypes.object,
    children: PropTypes.element.isRequired,
  }),
  mapProps(({ locale, messages, ...props }) => ({
    locale,
    key: locale,
    messages: messages[locale],
    ...props,
  })),
  IntlProvider);
