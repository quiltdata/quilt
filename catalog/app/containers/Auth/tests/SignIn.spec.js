import createHistory from 'history/createMemoryHistory';
import { fromJS } from 'immutable';

import LanguageProvider from 'containers/LanguageProvider';
import { Provider as APIProvider } from 'utils/APIConnector';
import { translationMessages as messages } from 'i18n';
import FormProvider from 'utils/ReduxFormProvider';
import StoreProvider from 'utils/StoreProvider';
import { nest } from 'utils/reactTools';
import configureStore from 'store';

import feature from 'testing/feature';
import errorSteps from 'testing/error';
import formSteps from 'testing/form';
import reactSteps from 'testing/react';
import requestsSteps from 'testing/requests';
import storageSteps from 'testing/storage';
import { mockComponentSelector } from 'testing/util';

import {
  Provider as AuthProvider,
  SignIn,
  errors,
} from '..';

import {
  api,
  tokensRaw,
  user,
  datasets,
} from './fixtures';

jest.mock('material-ui/RaisedButton');
jest.mock('material-ui/TextField');
jest.mock('react-router-dom');
jest.mock('components/Spinner');
jest.mock('constants/config');
jest.mock('utils/time');
jest.mock('utils/errorReporting');

const paths = {
  signIn: '/signin',
  page1: '/page1',
  page2: '/page2',
};
const signInRedirect = paths.page1;


const fields = ['password', 'username'];

const values = {
  'invalid password': 'invalid',
  'valid password': 'valid',
  username: 'bob',
};

const onSubmit = (ctx) => {
  const credentials = {
    username: ctx.mounted.find(`${mockComponentSelector('TextField')}[name="username"]`).prop('value'),
    password: ctx.mounted.find(`${mockComponentSelector('TextField')}[name="password"]`).prop('value'),
  };
  return { ...ctx, credentials };
};

const requests = {
  signIn: {
    setup: () => ['post', '/login'],
    expect: (ctx) =>
      expect.objectContaining({
        body: JSON.stringify(ctx.credentials),
      }),
    success: () => tokensRaw,
  },
  fetchUser: {
    setup: () => ['getOnce', '/me'],
    success: () => user,
  },
};

const setup = ({ storage, fetch = () => {}, next }) => {
  const history = createHistory({ initialEntries: ['/'] });
  const store = configureStore(fromJS({}), history);
  const search = next ? `?next=${encodeURIComponent(next)}` : '';
  const tree = nest(
    [StoreProvider, { store }],
    FormProvider,
    [LanguageProvider, { messages }],
    [APIProvider, { base: api, fetch }],
    [AuthProvider, { storage, signInRedirect }],
    [SignIn, { location: { search } }],
  );
  return { tree };
};

const steps = [
  ...requestsSteps(api, requests),
  ...formSteps({ values, fields, onSubmit }),
  ...errorSteps(errors),
  ...reactSteps({ setup }),
  ...storageSteps(datasets),
];


feature('containers/Auth/SignIn')
  .scenario('Redirecting when user is authenticated')

  .given('storage has current auth data')
  .given('next param is set to page2')

  .when('the component tree is mounted')
  .then('I should be redirected to page2')


  .scenario('Signing in')

  .given('storage has empty auth data')
  .given('signIn request is expected')
  .given('fetchUser request is expected')

  .when('the component tree is mounted')
  .then('I should see the form in interactive state')
  .then('the rendered markup should match the snapshot')

  .when('I submit the form')
  .then('I should see error on username field: "Enter your username"')
  .then('I should see error on password field: "Enter your password"')
  .then('I should see the form in invalid state')
  .then('the rendered markup should match the snapshot')

  .when('I enter username into username field')
  .then('I should see no error on username field')

  .when('I enter invalid password into password field')
  .then('I should see no error on password field')
  .then('I should see the form in valid state')

  .when('I submit the form')
  .then('I should see the form in waiting state')
  .then('the rendered markup should match the snapshot')

  .when('signIn request fails with 401')
  .then('I should see the form in invalid state')
  .then('I should see form error: "Invalid credentials"')
  .then('the rendered markup should match the snapshot')

  .when('I enter valid password into password field')
  .then('I should see no form error')
  .then('I should see the form in valid state')

  .when('I submit the form')
  .when('signIn request fails with 500')
  .then('I should see form error: "Something went wrong"')
  .then('I should see the form in invalid state')
  .then('the rendered markup should match the snapshot')
  .then('an AuthError error should be captured')

  .when('I re-enter valid password into password field')
  .then('I should see no form error')
  .then('I should see the form in valid state')

  .when('I submit the form')
  .when('signIn request succeeds')
  .when('fetchUser request succeeds')
  .then('I should be redirected to page1')

  .step(/I should be redirected to (.+)/, (ctx, path) => {
    ctx.mounted.update();
    expect(ctx.mounted.find(mockComponentSelector('Redirect')).prop('to'))
      .toBe(paths[path]);
  })

  .step(/next param is set to (.+)/, (ctx, path) => ({
    ...ctx,
    next: paths[path],
  }))

  .steps(steps)
  .run();
