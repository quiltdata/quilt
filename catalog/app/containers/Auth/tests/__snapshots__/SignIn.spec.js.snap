// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feature: containers/Auth/SignIn Background: Scenario: Signing in Given storage has empty auth data Given signIn request is expected Given fetchUser request is expected When the component tree is mounted Then the rendered markup should match the snapshot 1`] = `
.c0 {
  margin-left: auto;
  margin-right: auto;
  max-width: 280px;
  min-height: calc(100vh - 300px);
  width: 100%;
}

.c1 {
  text-align: center;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-top: 32px;
}

.c3 {
  font-size: 12px;
  line-height: 16px;
  margin-bottom: 12px;
  margin-top: 32px;
  text-align: center;
}

.c3 + .c3 {
  margin-top: 12px;
}

<div
  class="c0"
>
  <h1
    class="c1"
  >
    <span>
      Sign in
    </span>
  </h1>
  <form>
    <component-text-field
      disabled="false"
      fullwidth="true"
      name="username"
      value=""
    >
      <prop-floating-label-text>
        <span>
          Username
        </span>
      </prop-floating-label-text>
    </component-text-field>
    <component-text-field
      disabled="false"
      fullwidth="true"
      name="password"
      type="password"
      value=""
    >
      <prop-floating-label-text>
        <span>
          Password
        </span>
      </prop-floating-label-text>
    </component-text-field>
    <div
      class="c2"
    >
      <component-raised-button
        disabled="false"
        primary="true"
        type="submit"
      >
        <prop-label>
          <span>
            Sign in
          </span>
        </prop-label>
      </component-raised-button>
    </div>
    <p
      class="c3"
    >
      <span>
        Don't have an account? 
        <component-link
          to="/signup"
        >
          <span>
            Sign up
          </span>
        </component-link>
        .
      </span>
    </p>
    <p
      class="c3"
    >
      <span>
        Did you forget your password? 
        <component-link
          to="/reset_password"
        >
          <span>
            Reset it
          </span>
        </component-link>
        .
      </span>
    </p>
  </form>
</div>
`;

exports[`Feature: containers/Auth/SignIn Background: Scenario: Signing in Given storage has empty auth data Given signIn request is expected Given fetchUser request is expected When the component tree is mounted When I submit the form Then the rendered markup should match the snapshot 1`] = `
.c0 {
  margin-left: auto;
  margin-right: auto;
  max-width: 280px;
  min-height: calc(100vh - 300px);
  width: 100%;
}

.c1 {
  text-align: center;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-top: 32px;
}

.c3 {
  font-size: 12px;
  line-height: 16px;
  margin-bottom: 12px;
  margin-top: 32px;
  text-align: center;
}

.c3 + .c3 {
  margin-top: 12px;
}

<div
  class="c0"
>
  <h1
    class="c1"
  >
    <span>
      Sign in
    </span>
  </h1>
  <form>
    <component-text-field
      disabled="false"
      fullwidth="true"
      name="username"
      value=""
    >
      <prop-floating-label-text>
        <span>
          Username
        </span>
      </prop-floating-label-text>
      <prop-error-text>
        <span>
          Enter your username
        </span>
      </prop-error-text>
    </component-text-field>
    <component-text-field
      disabled="false"
      fullwidth="true"
      name="password"
      type="password"
      value=""
    >
      <prop-floating-label-text>
        <span>
          Password
        </span>
      </prop-floating-label-text>
      <prop-error-text>
        <span>
          Enter your password
        </span>
      </prop-error-text>
    </component-text-field>
    <div
      class="c2"
    >
      <component-raised-button
        disabled="true"
        primary="true"
        type="submit"
      >
        <prop-label>
          <span>
            Sign in
          </span>
        </prop-label>
      </component-raised-button>
    </div>
    <p
      class="c3"
    >
      <span>
        Don't have an account? 
        <component-link
          to="/signup"
        >
          <span>
            Sign up
          </span>
        </component-link>
        .
      </span>
    </p>
    <p
      class="c3"
    >
      <span>
        Did you forget your password? 
        <component-link
          to="/reset_password"
        >
          <span>
            Reset it
          </span>
        </component-link>
        .
      </span>
    </p>
  </form>
</div>
`;

exports[`Feature: containers/Auth/SignIn Background: Scenario: Signing in Given storage has empty auth data Given signIn request is expected Given fetchUser request is expected When the component tree is mounted When I submit the form When I enter username into username field When I enter invalid password into password field When I submit the form Then the rendered markup should match the snapshot 1`] = `
.c0 {
  margin-left: auto;
  margin-right: auto;
  max-width: 280px;
  min-height: calc(100vh - 300px);
  width: 100%;
}

.c1 {
  text-align: center;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-top: 32px;
}

.c3 {
  font-size: 12px;
  line-height: 16px;
  margin-bottom: 12px;
  margin-top: 32px;
  text-align: center;
}

.c3 + .c3 {
  margin-top: 12px;
}

<div
  class="c0"
>
  <h1
    class="c1"
  >
    <span>
      Sign in
    </span>
  </h1>
  <form>
    <component-text-field
      disabled="true"
      fullwidth="true"
      name="username"
      value="bob"
    >
      <prop-floating-label-text>
        <span>
          Username
        </span>
      </prop-floating-label-text>
    </component-text-field>
    <component-text-field
      disabled="true"
      fullwidth="true"
      name="password"
      type="password"
      value="invalid"
    >
      <prop-floating-label-text>
        <span>
          Password
        </span>
      </prop-floating-label-text>
    </component-text-field>
    <div
      class="c2"
    >
      <component-raised-button
        disabled="true"
        primary="true"
        type="submit"
      >
        <prop-label>
          <span>
            Sign in
          </span>
        </prop-label>
        <prop-children>
          <component-spinner
            style="font-size: 1.5em; opacity: 0.5; position: absolute; right: -1.5em;"
          />
        </prop-children>
      </component-raised-button>
    </div>
    <p
      class="c3"
    >
      <span>
        Don't have an account? 
        <component-link
          to="/signup"
        >
          <span>
            Sign up
          </span>
        </component-link>
        .
      </span>
    </p>
    <p
      class="c3"
    >
      <span>
        Did you forget your password? 
        <component-link
          to="/reset_password"
        >
          <span>
            Reset it
          </span>
        </component-link>
        .
      </span>
    </p>
  </form>
</div>
`;

exports[`Feature: containers/Auth/SignIn Background: Scenario: Signing in Given storage has empty auth data Given signIn request is expected Given fetchUser request is expected When the component tree is mounted When I submit the form When I enter username into username field When I enter invalid password into password field When I submit the form When signIn request fails with 401 Then the rendered markup should match the snapshot 1`] = `
.c0 {
  margin-left: auto;
  margin-right: auto;
  max-width: 280px;
  min-height: calc(100vh - 300px);
  width: 100%;
}

.c1 {
  text-align: center;
}

.c2 {
  color: #f44336;
  margin-top: 24px;
  text-align: center;
}

.c2 a {
  color: inherit !important;
  text-decoration: underline;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-top: 32px;
}

.c4 {
  font-size: 12px;
  line-height: 16px;
  margin-bottom: 12px;
  margin-top: 32px;
  text-align: center;
}

.c4 + .c4 {
  margin-top: 12px;
}

<div
  class="c0"
>
  <h1
    class="c1"
  >
    <span>
      Sign in
    </span>
  </h1>
  <form>
    <component-text-field
      disabled="false"
      fullwidth="true"
      name="username"
      value="bob"
    >
      <prop-floating-label-text>
        <span>
          Username
        </span>
      </prop-floating-label-text>
    </component-text-field>
    <component-text-field
      disabled="false"
      fullwidth="true"
      name="password"
      type="password"
      value="invalid"
    >
      <prop-floating-label-text>
        <span>
          Password
        </span>
      </prop-floating-label-text>
    </component-text-field>
    <p
      class="c2"
    >
      <span>
        Invalid credentials
      </span>
    </p>
    <div
      class="c3"
    >
      <component-raised-button
        disabled="true"
        primary="true"
        type="submit"
      >
        <prop-label>
          <span>
            Sign in
          </span>
        </prop-label>
      </component-raised-button>
    </div>
    <p
      class="c4"
    >
      <span>
        Don't have an account? 
        <component-link
          to="/signup"
        >
          <span>
            Sign up
          </span>
        </component-link>
        .
      </span>
    </p>
    <p
      class="c4"
    >
      <span>
        Did you forget your password? 
        <component-link
          to="/reset_password"
        >
          <span>
            Reset it
          </span>
        </component-link>
        .
      </span>
    </p>
  </form>
</div>
`;

exports[`Feature: containers/Auth/SignIn Background: Scenario: Signing in Given storage has empty auth data Given signIn request is expected Given fetchUser request is expected When the component tree is mounted When I submit the form When I enter username into username field When I enter invalid password into password field When I submit the form When signIn request fails with 401 When I enter valid password into password field When I submit the form When signIn request fails with 500 Then the rendered markup should match the snapshot 1`] = `
.c0 {
  margin-left: auto;
  margin-right: auto;
  max-width: 280px;
  min-height: calc(100vh - 300px);
  width: 100%;
}

.c1 {
  text-align: center;
}

.c2 {
  color: #f44336;
  margin-top: 24px;
  text-align: center;
}

.c2 a {
  color: inherit !important;
  text-decoration: underline;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-top: 32px;
}

.c4 {
  font-size: 12px;
  line-height: 16px;
  margin-bottom: 12px;
  margin-top: 32px;
  text-align: center;
}

.c4 + .c4 {
  margin-top: 12px;
}

<div
  class="c0"
>
  <h1
    class="c1"
  >
    <span>
      Sign in
    </span>
  </h1>
  <form>
    <component-text-field
      disabled="false"
      fullwidth="true"
      name="username"
      value="bob"
    >
      <prop-floating-label-text>
        <span>
          Username
        </span>
      </prop-floating-label-text>
    </component-text-field>
    <component-text-field
      disabled="false"
      fullwidth="true"
      name="password"
      type="password"
      value="valid"
    >
      <prop-floating-label-text>
        <span>
          Password
        </span>
      </prop-floating-label-text>
    </component-text-field>
    <p
      class="c2"
    >
      <span>
        Something went wrong. Try again later.
      </span>
    </p>
    <div
      class="c3"
    >
      <component-raised-button
        disabled="true"
        primary="true"
        type="submit"
      >
        <prop-label>
          <span>
            Sign in
          </span>
        </prop-label>
      </component-raised-button>
    </div>
    <p
      class="c4"
    >
      <span>
        Don't have an account? 
        <component-link
          to="/signup"
        >
          <span>
            Sign up
          </span>
        </component-link>
        .
      </span>
    </p>
    <p
      class="c4"
    >
      <span>
        Did you forget your password? 
        <component-link
          to="/reset_password"
        >
          <span>
            Reset it
          </span>
        </component-link>
        .
      </span>
    </p>
  </form>
</div>
`;
